### PR TITLE
Dqfd agorithm

### PR DESCRIPTION
**Pull request by Will-Nie**
Write dqfd algorithm. This commits include
- the algorithm policy
- the entry file (serial_entry_dqfd.py)
- the env configs (include cartpole, lunarlander, pong, space_invaders)

To run
- First generate your own expert model
- get the model path and add to the env_dqfd_config.py where env is one of the four
- run with the command `ding -m serial_dqfd -c env_dqfd_config.py -s 0` 

Todo:
- add test files for the algorithm
- add more env config
- add expert config instead of expert model

### CODE REVIEW

File: ding/entry/serial_entry_dqfd.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+    commander = BaseSerialCommander(
+        cfg.policy.other.commander, learner, collector, evaluator, replay_buffer, policy.command_mode
+    )
+    # ==========
+    # Main loop
+    # ==========
+    # Learner's before_run hook.
+    learner.call_hook('before_run')
+    expert_data = expert_collector.collect(n_sample=10000, policy_kwargs={'eps': -1})
+    for i in range(len(expert_data)):
```
Comment (PaParaZz1) [inline ding/entry/serial_entry_dqfd.py:109]: add comments in all related locations, such as `# set is_expert flag(expert 1, agent 0)`

File: ding/entry/serial_entry_dqfd.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+    expert_buffer.push(expert_data, cur_collector_envstep=0)
+    for _ in range(cfg.policy.learn.per_train_iter_k):
+        if evaluator.should_eval(learner.train_iter):
+            stop, reward = evaluator.eval(learner.save_checkpoint, learner.train_iter, collector.envstep)
+            if stop:
+                break
+        # Learn policy from collected data
+        # Expert_learner will train ``update_per_collect == 1`` times in one iteration.
+        train_data = expert_buffer.sample(learner.policy.get_attribute('batch_size'), learner.train_iter)
+        if train_data is None:
```
Comment (PaParaZz1) [inline ding/entry/serial_entry_dqfd.py:120]: expert_buffer is filled with enough data, so this check of whether `train_data is None` is not necessary.

File: ding/entry/serial_entry_dqfd.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+        new_data = collector.collect(train_iter=learner.train_iter, policy_kwargs=collect_kwargs)
+        for i in range(len(new_data)):
+            new_data[i]['is_expert'] = 0
+        replay_buffer.push(new_data, cur_collector_envstep=collector.envstep)
+        # Learn policy from collected data
+        for i in range(cfg.policy.learn.update_per_collect):
+            # Learner will train ``update_per_collect`` times in one iteration.
+            # The hyperparameter pho, the demo ratio, control the propotion of data coming\
+            # from expert demonstrations versus from the agent's own experience.
+            stats = torch.rand((learner.policy.get_attribute('batch_size'))) < cfg.policy.collect.pho
```
Comment (PaParaZz1) [inline ding/entry/serial_entry_dqfd.py:159]: `np.random.choice` can be more user friendly

File: ding/policy/dqfd.py [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+        self._target_model.update(self._learn_model.state_dict())
+        return {
+            'cur_lr': self._optimizer.defaults['lr'],
+            'total_loss': loss.item(),
+            'priority': td_error_per_sample.abs().tolist(),
+            # Only discrete action satisfying len(data['action'])==1 can return this and draw histogram on tensorboard.
+            # '[histogram]action_distribution': data['action'],
+        }
+
+    def _state_dict_learn(self) -> Dict[str, Any]:
```
Comment (PaParaZz1) [inline ding/policy/dqfd.py:218]: some methods like `_state_dict_learn` can extend `DQNPolicy` directly

File: ding/rl_utils/td.py [hunk 5]
```diff
@@ -329,6 +336,91 @@ def q_nstep_td_error(
...
+        target_q_s_a_one_step = nstep_return(
+            nstep_return_data(reward, target_q_s_a_one_step, done), gamma, nstep, value_gamma
+        )
+    td_error_one_step_per_sample = criterion(q_s_a, target_q_s_a_one_step.detach())
+
+    # calculate the supervised loss
+    device = q_s_a.device
+    max_action = torch.argmax(q, dim=-1)
+    Js = is_expert * (
+        q[batch_range, max_action.type(torch.int64)] +
```
Comment (PaParaZz1) [inline ding/rl_utils/td.py:416]: `max_action.long()`, `torch.where(action == max_action, torch.ones_like(action), torch.zeros_like(action))`

File: ding/rl_utils/td.py [hunk 6]
```diff
@@ -329,6 +336,91 @@ def q_nstep_td_error(
...
+    else:
+        target_q_s_a_one_step = nstep_return(
+            nstep_return_data(reward, target_q_s_a_one_step, done), gamma, nstep, value_gamma
+        )
+    td_error_one_step_per_sample = criterion(q_s_a, target_q_s_a_one_step.detach())
+
+    # calculate the supervised loss
+    device = q_s_a.device
+    max_action = torch.argmax(q, dim=-1)
+    Js = is_expert * (
```
Comment (PaParaZz1) [inline ding/rl_utils/td.py:415]: why use `Js` to name this variable

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/rl_utils/td.py [hunk 7]
```diff
@@ -329,6 +336,91 @@ def q_nstep_td_error(
...
+        q[batch_range, max_action.type(torch.int64)] +
+        0.8 * torch.from_numpy((action == max_action).numpy().astype(int)).float().to(device) - q_s_a
```
Comment (davide97l) [inline ding/rl_utils/td.py:417]: Could 0.8 be a parameter?

File: ding/entry/serial_entry_dqfd.py [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+        seed: int = 0,
+        env_setting: Optional[List[Any]] = None,
+        model: Optional[torch.nn.Module] = None,
+        expert_model: Optional[torch.nn.Module] = None,
+        max_iterations: Optional[int] = int(1e10),
+) -> 'Policy':  # noqa
+    """
+    Overview:
+        Serial pipeline sqil entry: we create this serial pipeline in order to\
+            implement SQIL in DI-engine. For now, we support the following envs\
```
[the remark this answers is not in the document]
Comment (Will-Nie) [reply ding/entry/serial_entry_dqfd.py:31]: Thx! corrected

[the remark this answers is not in the document]
Comment (Will-Nie) [reply ding/entry/serial_entry_dqfd.py:31]: Thx! corrected

File: ding/rl_utils/td.py (hunk 7 above)
Comment (Will-Nie) [reply ding/rl_utils/td.py:417]: Yes, the default margin loss is 0.8, but can be surely modified

File: ding/policy/dqfd.py [hunk 9]
```diff
@@ -0,0 +1,376 @@
...
+from .dqn import DQNPolicy
+from .common_utils import default_preprocess_learn
+from copy import deepcopy
+
+
+@POLICY_REGISTRY.register('dqfd')
+class DQFDPolicy(DQNPolicy):
+    r"""
+    Overview:
+        Policy class of DQN algorithm, extended by Double DQN/Dueling DQN/PER/multi-step TD.
```
[the remark this answers is not in the document]
Comment (Will-Nie) [reply ding/policy/dqfd.py:21]: Thx! corrected

File: ding/entry/serial_entry_dqfd.py [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+        policy=expert_policy.collect_mode,
+        tb_logger=tb_logger,
+        exp_name=cfg.exp_name
+    )
+    evaluator = InteractionSerialEvaluator(
+        cfg.policy.eval.evaluator, evaluator_env, policy.eval_mode, tb_logger, exp_name=cfg.exp_name
+    )
+    replay_buffer = create_buffer(cfg.policy.other.replay_buffer, tb_logger=tb_logger, exp_name=cfg.exp_name)
+    dummy_variable = deepcopy(cfg.policy.other.replay_buffer)
+    dummy_variable['replay_buffer_size'] = 10000
```
[the remark this answers is not in the document]
Comment (Will-Nie) [reply ding/entry/serial_entry_dqfd.py:98]: Yes for sure, I will make it a hyper-parameter

File: ding/entry/serial_entry_dqfd.py [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+    expert_buffer = create_buffer(dummy_variable, tb_logger=tb_logger, exp_name=cfg.exp_name)
```
Comment (puyuan1996) [inline ding/entry/serial_entry_dqfd.py:99]: where the following relative sampling is implemented?
> For proportional prior- itized sampling, different small positive constants, \epsilon_a and \epsilon_d, are added to the priorities of the agent and demonstration transitions to control the relative sampling of demonstration versus agent data

File: ding/policy/dqfd.py [hunk 12]
```diff
@@ -0,0 +1,376 @@
...
+
+        .. note::
+            We will vectorize ``process_transition`` and ``get_train_sample`` method in the following release version. \
+            And the user can customize the this data processing procecure by overriding this two methods and collector \
+            itself.
+        """
+        data_1 = deepcopy(get_nstep_return_data(data, 1, gamma=self._gamma))
+        data = get_nstep_return_data(data, self._nstep, gamma=self._gamma)
+        for i in range(len(data)):
+            data[i]['next_obs_1'] = data_1[i]['next_obs']
```
Comment (puyuan1996) [inline ding/policy/dqfd.py:305]: Here give some annotations could be better.

File: ding/rl_utils/td.py (hunk 6 above)
Comment (Will-Nie) [reply ding/rl_utils/td.py:415]: Yes, this is a bit confusing, I have modified it into JE which coincides with the original paper

File: ding/rl_utils/td.py (hunk 5 above)
Comment (Will-Nie) [reply ding/rl_utils/td.py:416]: Thx! corrected

File: ding/policy/dqfd.py (hunk 4 above)
Comment (Will-Nie) [reply ding/policy/dqfd.py:218]: Thx! these functions that can be inherited have been deleted

File: ding/entry/serial_entry_dqfd.py (hunk 3 above)
Comment (Will-Nie) [reply ding/entry/serial_entry_dqfd.py:159]: Thx! corrected.

File: ding/entry/serial_entry_dqfd.py (hunk 2 above)
Comment (Will-Nie) [reply ding/entry/serial_entry_dqfd.py:120]: Thx, I have double-checked and indeed, there is no need to add this part

File: ding/entry/serial_entry_dqfd.py (hunk 1 above)
Comment (Will-Nie) [reply ding/entry/serial_entry_dqfd.py:109]: Thx! I have added

File: ding/entry/serial_entry_dqfd.py (hunk 11 above)
Comment (Will-Nie) [reply ding/entry/serial_entry_dqfd.py:99]: we implemented the replay buffer based on the r2d3 paper. There are two separate buffer being used
the \epsilon_a and \epsilon_d can be replaced by the hyper-parameter pho

File: ding/policy/dqfd.py (hunk 12 above)
Comment (Will-Nie) [reply ding/policy/dqfd.py:305]: Thx! I have made some comments to make it clearer

File: ding/rl_utils/td.py (hunk 7 above)
Comment (Will-Nie) [reply ding/rl_utils/td.py:417]: I have made it a hyper-parameter
